druid/src/widget/aspect_ratio_box.rs

```diff
@@ -0,0 +1,279 @@
+            inner.lifecycle(ctx, event, data, env)
+        }
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, old_data, data, env)
+    )]
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &T, data: &T, env: &Env) {
+        if let Some(ref mut inner) = self.inner {
+            inner.update(ctx, old_data, data, env);
+        }
+    }
+
+    #[instrument(
+        name = "AspectRatioBox",
+        level = "trace",
+        skip(self, ctx, bc, data, env)
+    )]
+    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints, data: &T, env: &Env) -> Size {
+        bc.debug_check("AspectRatioBox");
+
+        let bc = if bc.max() == bc.min() {
+            warn!("Box constraints are tight. Aspect ratio box will not be able to preserve aspect ratio.");
+
+            *bc
+        } else if bc.max().width == f64::INFINITY && bc.max().height == f64::INFINITY {
+            warn!("Box constraints are INFINITE. Aspect ratio box won't be able to choose a size because the constraints given by the parent widget are INFINITE.");
+
+            // should I do this or should I just choose a default size
+            // I size the box with the child's size if there is one
+            let size = match self.inner.as_mut() {
+                Some(inner) => inner.layout(ctx, &bc, data, env),
+                None => Size::new(500., 500.),
+            };
+
+            BoxConstraints::new(Size::new(0., 0.), size)
+        } else {
+            let (mut box_width, mut box_height) = (bc.max().width, bc.max().height);
+
+            if self.ratio < 1.0 {
+                if (box_height >= box_width && box_height * self.ratio <= box_width)
+                    || box_width > box_height
+                {
+                    box_width = box_height * self.ratio;
+                } else if box_height >= box_width && box_height * self.ratio > box_width {
+                    box_height = box_width / self.ratio;
+                } else {
+                    // I'm not sure if these sections are necessary or correct
+                    // dbg!("ratio can't be preserved {}", self.ratio);
+                    warn!("The aspect ratio cannot be preserved because one of dimensions is tight and the other dimension is too small: bc.max() = {}, bc.min() = {}", bc.max(), bc.min());
+                }
+
+                BoxConstraints::tight(Size::new(box_width, box_height))
+            } else if self.ratio > 1.0 {
+                if box_width > box_height && box_height * self.ratio <= box_width {
+                    box_width = box_height * self.ratio;
+                }
+                // this condition might be wrong if box_width and height are equal to each other
+                // and the aspect ratio is something like 1.00000000000000001, in this case
+                // the  box_height * self.ratio could be equal to box_width
+                // however 1.00000000000000001 does equal 1.0
```

Comment:
this confuses me a bit. If you are not sure about the conditions, you can add `assert`s that state your assumptions, so we will at least find out when they are not being met.

Reply:
I mostly kept this here to see if you had any opinions on it. I don't think it will be an actual issue. I did test out those values to see if they would be a problem. I'm just paranoid with floating point numbers.